packages/@react-aria/grid/src/useGrid.ts

```diff
@@ -98,6 +98,41 @@ export function useGrid<T>(props: GridProps, state: GridState<T, GridCollection<
     scrollRef
   });
 
+  const cachedFocusedKey = useRef(null);
+  const cachedDelegate = useRef(null);
+  useEffect(() => {
+    let keyToFocus:Key;
+    if (
+      state.selectionManager.focusedKey === null &&
+      cachedFocusedKey.current !== null
+    ) {
+      let keyBelow = cachedDelegate.current.getKeyBelow(cachedFocusedKey.current);
+      while (keyBelow) {
+        if (state.collection.getItem(keyBelow)) {
+          keyToFocus = keyBelow;
+          break;
+        }
+        keyBelow = cachedDelegate.current.getKeyBelow(keyBelow);
```

Comment:
I still kinda feel like maybe we should base this on the order in the new version of the collection rather than the old one? For example, if a row was removed and the data was re-sorted in a single transaction, this could produce strange results:

Starting state:

* A
* B   <-- focused
* C
* D
* E
* F

Ending state:

* D
* E
* A
* F
* C  <-- focused

If instead, we based it on the index of the item in the old collection, and simply focused the corresponding item at that same index in the new collection, it would end up like this:

* D
* E   <-- focused
* A
* F
* C

This might be more natural to the user. Otherwise they might be scrolled around a lot from where they were.

Seems like there could be a lot of real world cases like this, e.g. changing a filter. A bunch of new items could appear at the same time as some are removed, leaving focus in a completely different place.

Additionally, if we base this on the index, we wouldn't need the keyboard delegate to find the next key and could move this logic into a lower level layer shared among all collections. Basically:

```js
let node = lastCollection.getItem(focusedKey);
let newItem = [...newCollection][node.index];
setFocusedKey(newItem.key);
```

Might also need to skip over disabled items (depending on the `disabledBehavior`) also if the corresponding item is disabled.

Reply:
We still need to worry about which column had focus in the row that was removed. Using the keyboard delegate makes this much simpler.

Reply:
Ah yeah that's true. Using the index does seem promising though in terms of the behavior, would just need to be two dimensional for tables.

Reply:
I'll investigate on Monday. Any thoughts on https://github.com/adobe/react-spectrum/pull/3885#discussion_r1063882356